infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java

```diff
@@ -143,102 +142,147 @@ public void checkpointState() {
             if (req.getTxnResolution()) {
-                // Then also need a read timestamp.
-                long timestamp = req.getReadTimestamp();
-                if (timestamp != -1L) {
-                    AtomicBoolean abort = new AtomicBoolean(false);
-                    for (UUID id : req.getStreams()) {
-                        if (abort.get())
-                            break;
-                        lastIssuedMap.compute(id, (k, v) -> {
-                            if (v == null) {
-                                return null;
-                            } else {
-                                if (v > timestamp) {
-                                    log.debug("Rejecting request due to {} > {} on stream {}", v, timestamp, id);
-                                    abort.set(true);
-                                }
-                            }
-                            return v;
-                        });
-                    }
-                    if (abort.get()) {
-                        globalIndex.getAndAdd(-req.getNumTokens());
-                        r.sendResponse(ctx, msg, CorfuMsgType.TOKEN_RES.payloadMsg(
-                                new TokenResponse(-1L, Collections.emptyMap(), Collections.emptyMap())));
-                        return;
-                    }
+                if (!txnResolution(req.getReadTimestamp(), req.getReadSet())) {
+                    // If the txn aborts, then DO NOT hand out a token.
+                    globalIndex.getAndAdd(-req.getNumTokens());
+                    r.sendResponse(ctx, msg, CorfuMsgType.TOKEN_RES.payloadMsg(
+                            new TokenResponse(-1L, Collections.emptyMap(), Collections.emptyMap())));
+                    return;
                 }
             }
 
-            ImmutableMap.Builder<UUID, Long> mb = ImmutableMap.builder();
-            ImmutableMap.Builder<UUID, Long> localAddresses = ImmutableMap.builder();
+            // If the txn can commit, or if the request is for a non-txn entry, then proceed normally to
+            // hand out local stream offsets.
+            ImmutableMap.Builder<UUID, Long> backPointerMap = ImmutableMap.builder();
+            ImmutableMap.Builder<UUID, Long> requestStreamTokens = ImmutableMap.builder();
             for (UUID id : req.getStreams()) {
-                lastIssuedMap.compute(id, (k, v) -> {
+                lastGlobalOffsetMap.compute(id, (k, v) -> {
                     if (v == null) {
-                        mb.put(k, -1L);
+                        backPointerMap.put(k, -1L);
                         return thisIssue + req.getNumTokens() - 1;
                     }
-                    mb.put(k, v);
+                    backPointerMap.put(k, v);
                     return Math.max(thisIssue + req.getNumTokens() - 1, v);
                 });
+                /*
+                 * Action table for (overwrite, replexOverwrite) pairs:
```

Comment:
Hmm ... I don't really know how to use Reviews in Github yet. I commented individually on your review comments in a previous review, but I don't know if they actually get submitted.. and/or if you get notified about them. Anyway, the description of overwrite/replexOverwrite is in TokenRequest.java. Should I move them or duplicate them in the Sequencer file? Basically, an overwrite is when an overwrite happens in the global log layer, and a replexOverwrite is when there is a successful global log write, but the write encounters an overwrite in the replex / stream layer.